### PR TITLE
[move-ir] Add phantom type parameters

### PR DESCRIPTION
**Pull request by nilehmann**
* Update syntax and compiler in move IR to support phantom parameters.
* Add tests for phantom type parameters.

## Motivation

Continue with the implementation of phantom parameters.

## Test Plan

`cargo xtest -p ir-testsuite phantom`

## Related PRs

This builds on top of https://github.com/diem/diem/pull/8599 which implements phantom types parameters in the bytecode.

### CODE REVIEW

File: language/testing-infra/module-generation/src/generator.rs [hunk 1]
```diff
@@ -335,3 +351,21 @@ impl<'a> ModuleGenerator<'a> {
         .gen()
     }
 }
+
+// Getter trait to be able to use [`StructTypeParameter`] and `(TypeVar, BTreeSet<Ability>)`
+// uniformily.
+trait TypeParamGetTypeVar {
```
Comment (nilehmann) [inline language/testing-infra/module-generation/src/generator.rs:357]: This is the same discussion as in the previous PR, luckily it is very local, and I thought it wouldn't be problematic to introduce the trait here because it's not public. I decided on the trait because the copy here is a bit more costly because `TypeVar`s are strings. The alternatives are:
* making `base_type` receive a `&[TypeVar]` which would require cloning all the type vars + the allocation of the array 
* making `base_type` receive a `&[&TypeVar]` which only require the extra array allocation

Comment (tnowacki) [reply language/testing-infra/module-generation/src/generator.rs:357]: I wouldn't be too worried about something like extra allocation, especially in this testing tool. I think it is better to stick to something less indirect, like some sort of `impl IntoIterator<Item = TypeVar>`

File: language/compiler/ir-to-bytecode/syntax/src/syntax.rs [hunk 2]
```diff
@@ -1209,6 +1209,19 @@ fn parse_type_var(tokens: &mut Lexer) -> Result<TypeVar, ParseError<Loc, anyhow:
     Ok(spanned(tokens.file_name(), start_loc, end_loc, type_var))
 }
 
+fn parse_type_parameter_with_phantom_decl(
+    tokens: &mut Lexer,
+) -> Result<(TypeVar, BTreeSet<Ability>, bool), ParseError<Loc, anyhow::Error>> {
+    let is_phantom = if tokens.peek() == Tok::Phantom {
```
Comment (tnowacki) [inline language/compiler/ir-to-bytecode/syntax/src/syntax.rs:1215]: I would probably make `phantom` a conditional keyword. So instead of making a new token, check for an identifier where its string value is "phantom"

File: language/move-lang/src/to_bytecode/translate.rs [hunk 3]
```diff
@@ -739,12 +727,24 @@ fn abilities(set: &AbilitySet) -> BTreeSet<IR::Ability> {
...
         .map(|tp| (type_var(tp.user_specified_name), abilities(&tp.abilities)))
         .collect()
 }
 
+fn struct_type_parameters(tps: Vec<TParam>) -> Vec<IR::StructTypeParameter> {
+    tps.into_iter()
+        .map(|tp| {
+            (
+                type_var(tp.user_specified_name),
+                abilities(&tp.abilities),
```
Comment (tnowacki) [inline language/move-lang/src/to_bytecode/translate.rs:741]: nit, I think this will be easier to read with some let bindings. So `let name = type_var(tp.user_specified_name); let constraints = abilities(&tp.abilities); (name, constraints, /* is_phantom */ false)`

File: language/move-ir/types/src/ast.rs [hunk 4]
```diff
@@ -229,15 +229,18 @@ pub type Fields<T> = Vec<(Field, T)>;
 #[derive(Clone, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
 pub struct StructName(String);
 
+/// A struct type parameter with its constraints and whether it's declared as phantom.
+pub type StructTypeParameter = (TypeVar, BTreeSet<Ability>, bool);
```
Comment (tnowacki) [inline language/move-ir/types/src/ast.rs:233]: nit, maybe move the bool to the front? to match the syntax

File: language/testing-infra/module-generation/src/generator.rs [hunk 5]
```diff
@@ -177,10 +177,29 @@ impl<'a> ModuleGenerator<'a> {
...
+            vec![]
+        } else {
+            let is_phantom = self.index(1) != 0;
+            let num_ty_params = self.index(self.options.max_ty_params);
+            let abilities = BTreeSet::from_iter(vec![Ability::Copy, Ability::Drop]);
+            init!(
+                num_ty_params,
+                (
+                    Spanned::unsafe_no_loc(TypeVar_::new(self.identifier())),
+                    abilities.clone(),
```
Comment (tnowacki) [inline language/testing-infra/module-generation/src/generator.rs:192]: same as above, I'd try to get all of the tuple args on one line, I find it a bit easier to read

**Review by tnowacki — COMMENTED** [review]
Overall things are looking good! Will just need some tests

File: language/move-lang/src/to_bytecode/translate.rs [hunk 6]
```diff
@@ -739,12 +727,22 @@ fn abilities(set: &AbilitySet) -> BTreeSet<IR::Ability> {
...
         .map(|tp| (type_var(tp.user_specified_name), abilities(&tp.abilities)))
         .collect()
 }
 
+fn struct_type_parameters(tps: Vec<TParam>) -> Vec<IR::StructTypeParameter> {
+    tps.into_iter()
+        .map(|tp| {
+            let name = type_var(tp.user_specified_name);
+            let constraints = abilities(&tp.abilities);
+            (/* is_phantom*/ false, name, constraints)
```
Comment (tnowacki) [inline language/move-lang/src/to_bytecode/translate.rs:741]: ultra nit spacing between `is_phantom` and `*/` :P

File: language/testing-infra/module-generation/src/generator.rs [hunk 7]
```diff
@@ -160,27 +165,39 @@ impl<'a> ModuleGenerator<'a> {
...
             let abilities = BTreeSet::from_iter(vec![Ability::Copy, Ability::Drop]);
-            init!(
-                num_ty_params,
-                (
-                    Spanned::unsafe_no_loc(TypeVar_::new(self.identifier())),
-                    abilities.clone(),
-                )
-            )
+            init!(num_ty_params, {
+                let name = Spanned::unsafe_no_loc(TypeVar_::new(self.identifier()));
```
Comment (tnowacki) [inline language/testing-infra/module-generation/src/generator.rs:176]: bit weird to nest the name like this unless you have to. Might look better as 
```
let name = Spanned....
init!(num_ty_params, (name, abilities.clone()))
```

File: language/testing-infra/module-generation/src/generator.rs [hunk 8]
```diff
@@ -160,27 +165,39 @@ impl<'a> ModuleGenerator<'a> {
...
+    fn struct_type_parameters(&mut self) -> Vec<StructTypeParameter> {
+        // Don't generate type parameters if we're generating simple types only
+        if self.options.simple_types_only {
+            vec![]
+        } else {
+            let is_phantom = self.index(1) != 0;
+            let num_ty_params = self.index(self.options.max_ty_params);
+            let abilities = BTreeSet::from_iter(vec![Ability::Copy, Ability::Drop]);
+            init!(num_ty_params, {
+                let name = Spanned::unsafe_no_loc(TypeVar_::new(self.identifier()));
```
Comment (tnowacki) [inline language/testing-infra/module-generation/src/generator.rs:191]: same as above

File: language/ir-testsuite/tests/move/generics/phantom_params/struct_definition_ok.mvir [hunk 9]
```diff
@@ -0,0 +1,42 @@
+module M1 {
+    // Not using a phantom parameter at all is ok.
+    struct S1<phantom T> {
+        a: u64
+    }
+
+    // Phantom parameters can be used in phantom position.
+    struct S2<phantom T> {
+        a: Self.S1<T>,
+        b: vector<Self.S1<T>>
```
Comment (tnowacki) [inline language/ir-testsuite/tests/move/generics/phantom_params/struct_definition_ok.mvir:10]: nice!

File: language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    struct S1<T1: drop, T2: copy, T3: store, T4: key> { a: bool }
+    struct S2 {
+        a: Self.S1<M1.HasDrop<M1.NoAbilities>, M1.HasCopy<M1.NoAbilities>, M1.HasStore<M1.NoAbilities>, M1.HasKey<M1.NoAbilities>>,
+    }
+
+    f1<T1: drop, T2: copy, T3: store, T4: key>() { return; }
+    f2() {
+        Self.f1<M1.HasDrop<M1.NoAbilities>, M1.HasCopy<M1.NoAbilities>, M1.HasStore<M1.NoAbilities>, M1.HasKey<M1.NoAbilities>>();
+        return;
+    }
```
Comment (tnowacki) [inline language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir:84]: The constraint behavior here feels a bit different than the bytecode related checks above. Maybe split it out into a separate test?

File: language/ir-testsuite/tests/move/generics/phantom_params/struct_definition_ok.mvir [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+//! new-transaction
+
+module M3 {
+    // Phantom parameters should be allowed to be declared with constraints.
+
+    struct S1<phantom T: copy> {
+        a: u64
+    }
+
+    struct S2<phantom T: copy> {
```
Comment (tnowacki) [inline language/ir-testsuite/tests/move/generics/phantom_params/struct_definition_ok.mvir:39]: maybe have more than one constraint here, e.g. `copy + drop + store`

File: language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir [hunk 12]
```diff
@@ -0,0 +1,85 @@
+// This file contains tests checking that type arguments used in phantom position
```
Comment (tnowacki) [inline language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir:1]: might be good to have negative versions of all of these tests.... thought I can't really see what those would be?

Maybe have some tests like `HasAbility<phantom T1, T2> has Ability` where `T2` does and does not have that ability. To show `T1` doesn't impact it. Thoughts?

**Review by tnowacki — COMMENTED** [review]
Overall looks good! Just some final touches on tests and it should be good to go

File: language/testing-infra/module-generation/src/generator.rs (hunk 7 above)
Comment (nilehmann) [reply language/testing-infra/module-generation/src/generator.rs:176]: you have to because `self.identifier()` has side effects and you want to call it in each iteration.

File: language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir (hunk 10 above)
Comment (nilehmann) [reply language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir:84]: for me is a variation of the same. you are checking that a struct instantiation has the correct abilities by using them in different places that require them to have those abilities.

Comment (tnowacki) [reply language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir:84]: Yeah but the actual code that you are testing is in different locations... You are right that they are conceptually the same, but tests are more helpful the smaller they are, since when something breaks it is easier to debug

File: language/testing-infra/module-generation/src/generator.rs (hunk 7 above)
Comment (tnowacki) [reply language/testing-infra/module-generation/src/generator.rs:176]: oof. Alright I guess keep it as is then

File: language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir (hunk 12 above)
Comment (nilehmann) [reply language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir:1]: The negative test feels weird to me and I don't think it adds much here as not having the ability doesn't actually mean `T1` doesn't have an impact. Also, "negative bugs" are not too bad because fixing them is always going to be backward-compatible as you would be lifting restrictions. This may be the reason they feel weird to me.

Mixing phantom and non-phantom parameters make a lot of sense though. I'll start by adding that to the existing tests.

File: language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir (hunk 10 above)
Comment (nilehmann) [reply language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir:84]: ok, makes sense. I put it into a separate file.

File: language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir (hunk 12 above)
Comment (nilehmann) [reply language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir]: well, I end up doing anyway...

Comment (tnowacki) [reply language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir]: Here I was referring to negative tests as tests that throw errors. But it is also very important to test any sort of "negative bugs" resulting from a lack of precision in the analysis. The tests are more there to catch some sort of inconsistency that arises from a change, or to track features as they develop. In short, if they start passing, be afraid :P 

I'll take a look at the test changes

File: language/ir-testsuite/tests/move/generics/phantom_params/bytecode_ops_abilities_bad.mvir [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+
+    // `CopyLoc` requires copy
+    f(x: Self.HasCopy<M1.NoAbilities, M1.NoAbilities>): Self.HasCopy<M1.NoAbilities, M1.NoAbilities> * Self.HasCopy<M1.NoAbilities, M1.NoAbilities> {
+        return (copy(x), move(x));
+    }
+ }
+
+ //! new-transaction
+
+ module M6 {
```
Comment (tnowacki) [inline language/ir-testsuite/tests/move/generics/phantom_params/bytecode_ops_abilities_bad.mvir:63]: I don't see any with store. You should do something like `struct Cup<T: store> has key { value: T }` and `Cup<HasStore<NoAbilities>>` or even just `K has key { value: HasStore<NoAbilities>> }`

The generic one might be good since it will be checked at the operation, instead of the declaration

File: language/ir-testsuite/tests/move/generics/phantom_params/bytecode_ops_abilities_ok.mvir [hunk 14]
```diff
@@ -0,0 +1,48 @@
+// Test checking that type arguments used in phantom position (phantom arguments)
+// are not considered when deriving the abilities for a struct by checking against
+// abilities required for specific bytecode operatiosn.
+
+module M {
+    struct NoAbilities { a: bool }
+    struct HasDrop<phantom T1, T2> has drop { a: bool }
+    struct HasCopy<phantom T1, T2> has copy { a : bool }
+    struct HasKey<phantom T1, T2> has key { a : bool }
```
Comment (tnowacki) [inline language/ir-testsuite/tests/move/generics/phantom_params/bytecode_ops_abilities_ok.mvir:9]: Same as above, we should have store tests

File: language/ir-testsuite/tests/move/generics/phantom_params/constraints_abilities_bad.mvir [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+    struct NoAbilities { a: bool }
+    struct HasAbilities<phantom T1, T2> has drop, copy, store, key { a: bool }
+}
+
+//! new-transaction
+
+module M2 {
+    import {{default}}.M1;
+
+    struct S1<T: drop + copy + store + key> { a: bool }
```
Comment (tnowacki) [inline language/ir-testsuite/tests/move/generics/phantom_params/constraints_abilities_bad.mvir:11]: There should also be a test for a single constraint

File: language/ir-testsuite/tests/move/generics/phantom_params/constraints_abilities_ok.mvir [hunk 16]
```diff
@@ -0,0 +1,15 @@
+module M {
+    struct NoAbilities { a: bool }
+    struct HasAbilities<phantom T1, T2> has drop, copy, store, key { a: bool }
+
+    struct S1<T: drop + copy + store + key> { a: bool }
```
Comment (tnowacki) [inline language/ir-testsuite/tests/move/generics/phantom_params/constraints_abilities_ok.mvir:5]: same as above, good to have the single and multiple cases

**Review by tnowacki — COMMENTED** [review]
Mostly good, just a last touch on tests and should be good to go! 

Luckily you can then mostly re-use these tests for the source language :) (but will be easier without all the `//! new-transaction` stuff and probably just in a single module)

File: language/ir-testsuite/tests/move/generics/phantom_params/bytecode_ops_abilities_bad.mvir (hunk 13 above)
Comment (nilehmann) [reply language/ir-testsuite/tests/move/generics/phantom_params/bytecode_ops_abilities_bad.mvir:63]: Both tests are already considered. The first one in `abilities_constraints_[ok/bad].mvir` and the second in `fields_abilities_[ok/bad].mvir`.

File: language/ir-testsuite/tests/move/generics/phantom_params/constraints_abilities_bad.mvir (hunk 15 above)
Comment (nilehmann) [reply language/ir-testsuite/tests/move/generics/phantom_params/constraints_abilities_bad.mvir:11]: done

File: language/ir-testsuite/tests/move/generics/phantom_params/constraints_abilities_ok.mvir (hunk 16 above)
Comment (nilehmann) [reply language/ir-testsuite/tests/move/generics/phantom_params/constraints_abilities_ok.mvir:9]: done

File: language/ir-testsuite/tests/move/generics/phantom_params/bytecode_ops_abilities_bad.mvir (hunk 13 above)
Comment (tnowacki) [reply language/ir-testsuite/tests/move/generics/phantom_params/bytecode_ops_abilities_bad.mvir:63]: That is not quite the same test though, as it is not testing the interaction at the bytecode operation itself

Comment (nilehmann) [reply language/ir-testsuite/tests/move/generics/phantom_params/bytecode_ops_abilities_bad.mvir:63]: Ok, I added the test.

**Review by tnowacki — DISMISSED** [review]

**Comment by tnowacki** [comment]
@nilehmann let me know if this is good to go, and I will land it. Thanks for all the iterations on tests!

**Comment by nilehmann** [comment]
@tnowacki, yes, it's good to go!
